org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/cleanup/CleanUpsTest.java

```diff
@@ -240,6 +244,235 @@ public void testInvertEqualsCleanUp() throws Exception {
+				+ "			System.out.println(real.substring(0));\n" //
+				+ "		}\n" //
+				+ "	}\n" //
+				+ "}";
+
+		assertEquals(expected, actual);
+	}
+
+	@Test
+	public void testLambdaExpression() throws Exception {
+		String contents = "package test1;\n" //
+				+ "\n" //
+				+ "import java.util.function.IntConsumer;\n" //
+				+ "\n" //
+				+ "public class LambdaExpression {\n" //
+				+ "    public void test() {\n" //
+				+ "        IntConsumer c = new IntConsumer() {\n" //
+				+ "            @Override\n" //
+				+ "            public void accept(int value) {\n" //
+				+ "                System.out.println(value);\n" //
+				+ "	           }\n" //
+				+ "        };\n" //
+				+ "    }\n" //
+				+ "}";
+
+		ICompilationUnit unit = pack1.createCompilationUnit("LambdaExpression.java", contents, false, monitor);
+		String uri = unit.getUnderlyingResource().getLocationURI().toString();
+		List<TextEdit> textEdits = registry.getEditsForAllActiveCleanUps(new TextDocumentIdentifier(uri), Arrays.asList("lambdaExpression"), monitor);
+		String actual = TextEditUtil.apply(unit, textEdits);
+		String expected = "package test1;\n" //
+				+ "\n" //
+				+ "import java.util.function.IntConsumer;\n" //
+				+ "\n" //
+				+ "public class LambdaExpression {\n" //
+				+ "    public void test() {\n" //
+				+ "        IntConsumer c = value -> System.out.println(value);\n" //
+				+ "    }\n" //
+				+ "}";
+
+		assertEquals(expected, actual);
+	}
+
+	@Test
+	public void testMultiCleanup() throws Exception {
+		String contents = "package test1;\n"
+				+ "\n"
+				+ "import java.io.File;\n"
+				+ "import java.io.FileFilter;\n"
+				+ "import java.util.Arrays;\n"
+				+ "\n"
+				+ "public class MutliCleanup {\n"
+				+ "    public void test() {\n"
+				+ "        String PATH = \"/this/is/some/path\";\n"
+				+ "        String MESSAGE = \"This is a message.\" +\n"
+				+ "                \"This message has multiple lines.\" +\n"
+				+ "                \"We can convert it to a text block\";\n"
+				+ "\n"
+				+ "        Object[] obj = Arrays.asList(PATH).toArray();\n"
+				+ "        if (obj[0] instanceof String) {\n"
+				+ "            String tmp = (String) obj[0];\n"
+				+ "            File f = new File(tmp);\n"
+				+ "            File[] filtered = f.listFiles(new FileFilter() {\n"
+				+ "                @Override\n"
+				+ "                public boolean accept(File path) {\n"
+				+ "                    return true;\n"
+				+ "                }\n"
+				+ "            });\n"
+				+ "        }\n"
+				+ "    }\n"
+				+ "}\n"
+				+ "";
+
+		ICompilationUnit unit = pack1.createCompilationUnit("MultiCleanup.java", contents, false, monitor);
+		String uri = unit.getUnderlyingResource().getLocationURI().toString();
+		List<TextEdit> textEdits = registry.getEditsForAllActiveCleanUps(new TextDocumentIdentifier(uri), Arrays.asList("lambdaExpression", "instanceofPatternMatch", "stringConcatToTextBlock"), monitor);
```

Comment:
@datho7561 

Currently, this test works. However, if you add `addFinalModifier` to the list, I get a failure in TextEditUtil. Initially I tried this in a client (eg. vscode-java), and the text edits get returned to the client, but nothing happens. So this isn't just TextEditUtil, but an underlying issue with the edits.

Either a bug in `addFinalModifier` or some kind of issue with the cleanup registry calculation.

Reply:
Hmm. I'll double check this. When I tried `addFinalModifier` this morning, it seemed to be working.

Reply:
It looks like it's based around the assumption I made for the clean ups that the changes they propose wouldn't overlap. For instance, the final modifier change wants to put `final` on the `File` in the anonymous class, but that overlaps with the lambda clean up. Also, the final modifier makes one big text edit that changes all the text from its first change to the last change, so it overlaps in other spots.

We will probably need a way to manage applying the clean up to a version of the compilation unit with the changes from the previous clean ups applied, then calculate how to perform all the changes at once without using overlapping text edits.

Reply:
Right, I just looked at the clean up loop. It was never going to work too well with overlaps, particularly if lines get adjusted. Like you suggest, we just need to apply the current text edit to a temporary document/buffer/cu and use the new document in the context for the next cleanup.

Or, if for some reason we can't guarantee clients will respect the order of text edits (shouldn't be the case) just compute the resulting document, and if there were a helper method to compute a text edit given a before/after doc (this is basically edit distance).